src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Controller_ReturnKey.cs

```diff
@@ -64,7 +65,8 @@ private void CommitOnEnter(out bool sendThrough, out bool committed)
             // to the editor.  In single-line debugger windows (Watch, etc), however, we don't
             // want to send the enter though, because those windows don't support displaying
             // more than one line of text.
-            sendThrough = !model.TriggerInfo.IsDebugger || model.TriggerInfo.IsImmediateWindow;
+            sendThrough = !model.Trigger.CustomTags.Contains(WellKnownCompletionTriggerTags.Debugger)
```

Comment:
Are there any custom triggers being used to represent other types of triggering, or is it really being used as "here's a way to pass ambient state along for the ride"? I recognize this might churn stuff, but it very much feels like this CustomTags exists on trigger just because it was the convenient type that happens to be there.

Three interesting, potentially controversial, and likely terrible ideas came to mind while writing that previous sentence:

1. I think you already had some sort of "context" object that was beginning to congeal. Perhaps a type that is the "context" that contains the trigger, various "environment" flags (notably this debugger flag and the interactive flag), and maybe some other pieces of information could become the replacement.
2. Is the debugger flag only being used to control whether we list completion items before their declaration point? And do completion providers get passed an OptionSet from the originating workspace (or can get at it)? If both are "yes", then _perhaps_ we should instead have there be a not-UI-exposed completion option that simply specifies that debugger behavior, and the debugger workspace we create just sets that option explicitly. I could _imagine_ people might actually like that behavior during real typing as well. It's not rare that you might write some code and realize you need to use a local declared farther down. Rather than me breaking context, moving the local up, and then writing the code again, that it's perhaps OK to actually complete it anyways. It's I am _not_ saying we should expose it in a UI (at least not until customers come with pitchforks asking for it :smile:), but if they did come along and ask for that we'd implement that with an option. Given that *also* might fix this problem too, I can't help but notice the coincidence.
3. Rather than having some additional way to shuttle info, this could just be checking the workspace kind. It's still comparing magic strings, but at least without the baggage of extra types.